### PR TITLE
Collapse IMediator into one interface

### PR DESCRIPTION
**Pull request by jbogard**
No caching of results - yet. I'd build a cache for an action, not a type though.

### CODE REVIEW

**Comment by jbogard** [comment]
@khellang i didn't get rid of the other interfaces :)

File: samples/MediatR.Examples/PingAsyncHandler.cs [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
     public class PingAsyncHandler : IAsyncRequestHandler<PingAsync, Pong>
     {
-        public async Task<Pong> Handle(PingAsync message)
+        public Task<Pong> Handle(PingAsync message)
         {
-            return await Task.Factory.StartNew(() => new Pong { Message = message.Message + " Pong" });
+            return Task.Factory.StartNew(() => new Pong { Message = message.Message + " Pong" });
```
Comment (khellang) [inline samples/MediatR.Examples/PingAsyncHandler.cs]: `Task.FromResult`?

File: src/MediatR/Mediator.cs [hunk 2]
```diff
@@ -31,193 +30,122 @@ public Mediator(SingleInstanceFactory singleInstanceFactory, MultiInstanceFactor
...
-            var pipeline = GetPipeline(request, invokeHandler);
-
-            return pipeline;
+                var wrapper =
+                    (RequestHandlerWrapper)
+                    Activator.CreateInstance(typeof(RequestHandlerWrapperImpl<>).MakeGenericType(requestType), handler);
+                return () =>
+                {
+                    wrapper.Handle(request);
+                    return Task.FromResult(Unit.Value);
```
Comment (khellang) [inline src/MediatR/Mediator.cs]: Can we cache this? `Unit.Task`?

File: src/MediatR/Mediator.cs [hunk 3]
```diff
@@ -31,193 +30,122 @@ public Mediator(SingleInstanceFactory singleInstanceFactory, MultiInstanceFactor
...
-
-        private TWrapper GetHandler<TWrapper, TResponse>(object request, Type handlerType, Type wrapperType)
-        {
-            var requestType = request.GetType();
-
-            var genericHandlerType = _genericHandlerCache.GetOrAdd(handlerType, new ConcurrentDictionary<Type, Type>())
-                .GetOrAdd(requestType, handlerType, (type, root) => root.MakeGenericType(type, typeof(TResponse)));
-            var genericWrapperType = _wrapperHandlerCache.GetOrAdd(wrapperType, new ConcurrentDictionary<Type, Type>())
-                .GetOrAdd(requestType, wrapperType, (type, root) => root.MakeGenericType(type, typeof(TResponse)));
+                .Select(handler => Task.Run(() => handler.Handle(notification), cancellationToken));
```
Comment (khellang) [inline src/MediatR/Mediator.cs]: Was this `Task.Run` intentional? I think we should steer as far away from `Task.Run` as possible.

File: src/MediatR/Mediator.cs (hunk 2 above)
Comment (jbogard) [inline src/MediatR/Mediator.cs]: Good call.

File: src/MediatR/Mediator.cs (hunk 3 above)
Comment (jbogard) [inline src/MediatR/Mediator.cs]: Another option is to execute them synchronously I guess? That's probably better, but I needed a hot task.

Comment (khellang) [inline src/MediatR/Mediator.cs]: `Task.FromResult` is always the best option to wrap a synchronous result in a `Task` :smile:

Comment (khellang) [inline src/MediatR/Mediator.cs]: Or even better; `ValueTask`, but not sure how adoptable that is yet...

**Pull request merged**